### PR TITLE
Refactored CrossEncoder into our own wrapper class to support head training

### PR DESCRIPTION
**Pull request by Dmitryv-2024**
Сейчас в качестве классификатора используем LogisticRegressionCV из sklearn (как раньше). Можно сделать несколько улучшений:
- LogisticRegressionCV поддерживает class_weight="balanced". Может стоит убрать balance_factor из construct_samples и положится на реализацию sklearn для сильно разбалансированных меток?
- Можно реализовать тренировку своего классификатора. Например: сделать два полносвязных слоя и использовать label smoothing + kl_div loss function (идея из кода к статье DNCC, но у них в классификаторе был один слой, поэтому предлагаю попробовать два слоя, т.к. не уверен что имеет смысл что-то делать с одним слоем, вряд-ли будет лучше логистической регрессии)

В классе Dataset добавил ключ - не разбивать train (по умолчанию делим на два как и раньше), т.к. для тестов нужно чуть больше данных.

### CODE REVIEW

**Comment by Samoed** [comment]
А почему у нас тут используется предсказания из CrossEncoder, а не промто эмбеды? Кажется что делать классификацию на предсказаниях кросс энкодера так себе

**Comment by Dmitryv-2024** [comment]
> А почему у нас тут используется предсказания из CrossEncoder, а не промто эмбеды? Кажется что делать классификацию на предсказаниях кросс энкодера так себе

Реализованы два режима:
- простая обертка над CrossEncoder - используем предсказания кросс энкодера
- С тренировкой классификатора - используем эмбединги из кроссэнкодера (hook установлен на слой классификатора в метод _classifier_hook) и тренируем логистическую регрессию.

**Comment by Samoed** [comment]
Я говорю про то чтобы не использовать _classifier_hook можно просто сделать SentenceTranformer, а не CrossEncoder

**Comment by Dmitryv-2024** [comment]
> Я говорю про то чтобы не использовать _classifier_hook можно просто сделать SentenceTranformer, а не CrossEncoder

Нам нужно проверять сходство двух предложений. CrossEncoder как раз на это и натренирован, т.е. ожидаем, что его эмбединги хороши для нашей задачи. Есть гипотеза, что спомощью тренировки только головы, мы можем получить дешевый finetuning.

File: autointent/_transformers/_nli_transformer.py [hunk 1]
```diff
@@ -87,18 +87,41 @@ class CrossEncoderWithLogreg:
...
-        Initialize the CrossEncoderWithLogreg.
-
-        :param model: The CrossEncoder model to use.
+    def __init__(
+        self,
+        model: str,
+        device: str = "cpu",
+        train_classifier: bool = False,
+        batch_size: int = 326,
+        max_length: int | None = None,
```
Comment (voorhs) [inline autointent/_transformers/_nli_transformer.py:99]: а этот параметр потом где-то используется?

File: autointent/_transformers/_nli_transformer.py [hunk 2]
```diff
@@ -87,18 +87,41 @@ class CrossEncoderWithLogreg:
...
+        if classifier_head is not None or train_classifier:
+            self.train_classifier = True
+            self._logits_list: list[npt.NDArray[Any]] = []
+            self._hook_handler = self.cross_encoder.model.classifier.register_forward_hook(self._classifier_hook)
+
+    def _classifier_hook(self, _module, input_tensor, _output_tensor) -> None:  # type: ignore[no-untyped-def] # noqa: ANN001
+        self._logits_list.append(input_tensor[0].cpu().numpy())
 
     @torch.no_grad()
     def get_features(self, pairs: list[list[str]]) -> npt.NDArray[Any]:
```
Comment (voorhs) [inline autointent/_transformers/_nli_transformer.py]: надо бы в докстринге и/или в названии функции отразить что эта функция возвращает либо сырые активации, либо уже готовые предсказания в завиимости от значения `train_classifier`

File: autointent/_transformers/_nli_transformer.py [hunk 3]
```diff
@@ -87,18 +87,41 @@ class CrossEncoderWithLogreg:
...
+        self.max_length = max_length
+        self._clf = classifier_head
+
+        if classifier_head is not None or train_classifier:
+            self.train_classifier = True
+            self._logits_list: list[npt.NDArray[Any]] = []
+            self._hook_handler = self.cross_encoder.model.classifier.register_forward_hook(self._classifier_hook)
+
+    def _classifier_hook(self, _module, input_tensor, _output_tensor) -> None:  # type: ignore[no-untyped-def] # noqa: ANN001
+        self._logits_list.append(input_tensor[0].cpu().numpy())
```
Comment (voorhs) [inline autointent/_transformers/_nli_transformer.py]: вижу свою вину в неправильном названии переменной `logits_list`, т.к. изначально это я так придумал в классе `CrossEncoderWithLogreg`

правильнее назвать это `activations_list`, логитами в мл называют другое)

File: autointent/_transformers/_nli_transformer.py (hunk 2 above)
Comment (voorhs) [reply autointent/_transformers/_nli_transformer.py]: еще наверное надо подправить аннотацию типа для `pairs`: правильнее `list[tuple[str,str]]`

File: autointent/_transformers/_nli_transformer.py [hunk 4]
```diff
@@ -108,20 +131,15 @@ def get_features(self, pairs: list[list[str]]) -> npt.NDArray[Any]:
...
-            self.cross_encoder.predict(batch)
-
-        handler.remove()
-
-        return np.concatenate(logits_list, axis=0)
+        res = self._logits_list
+        self._logits_list = []
+        return np.concatenate(res, axis=0)
 
     def _fit(self, pairs: list[list[str]], labels: list[LabelType]) -> None:
```
Comment (voorhs) [inline autointent/_transformers/_nli_transformer.py]: и здесь аннотация типа у `pairs`

File: autointent/_transformers/_nli_transformer.py [hunk 5]
```diff
@@ -161,8 +184,40 @@ def predict(self, pairs: list[list[str]]) -> npt.NDArray[Any]:
         :param pairs: List of text pairs to classify.
```
Comment (voorhs) [inline autointent/_transformers/_nli_transformer.py:187]: и тут аннотация типа у `pairs`

File: autointent/modules/scoring/_dnnc/dnnc.py [hunk 6]
```diff
@@ -192,8 +190,6 @@ def fit(self, utterances: list[str], labels: list[LabelType]) -> None:
         """
         self.n_classes = len(set(labels))
 
-        self.model = CrossEncoder(self.cross_encoder_name, trust_remote_code=True, device=self.device)
-
         vector_index_client = VectorIndexClient(self.device, self.db_dir, embedder_use_cache=self.embedder_use_cache)
 
         if self.prebuilt_index:
```
Comment (voorhs) [inline autointent/modules/scoring/_dnnc/dnnc.py]: вот эта логика, как и многое что написано в `from_context`, уже устарела, я сам исправлю это

**Comment by voorhs** [comment]
> Я говорю про то чтобы не использовать _classifier_hook можно просто сделать SentenceTranformer, а не CrossEncoder

причины почему мы используем хук:
- это работает :)
- нам нужно использовать один трансформер в двух режимах в зависимости от флага `train_classifier`: с претренированной головой и со своей кастомной

Претренированную голову точно можно ипользовать с помощью `CrossEncoder.predict()`, а о том что ее можно использовать и с помощью какого-то метода класса `SentenceTransformer` мне неизвестно. Если это и можно сделать, то наверное это что-то экзотическое и не задокументировано в самой библиотеке `sentence_transformers`

**Review by voorhs — COMMENTED** [review]
мне неожиданно понадобился этот PR поэтому я сделал ревью сейчас))

в целом все отлично, есть только один странный момент: кажется внутри конструктора `NLITransformer` сохраняется параметр `max_length` но никак не используется

поскольку мне сейчас очень нужен этот PR, я сам готов исправить все замечания и разрешить все конфликты (планирую это сделать в ближайшие пару часов)

UPD еще есть вопрос по поводу `split_train`: где именно возникала ошибка? просто сейчас если судить по тестам как будто все равно везде стоит `split_train=True` и ничего не падает

**Comment by Samoed** [comment]
> > Я говорю про то чтобы не использовать _classifier_hook можно просто сделать SentenceTranformer, а не CrossEncoder
> 
> причины почему мы используем хук:
> 
>     * это работает :)
> 
>     * нам нужно использовать один трансформер в двух режимах в зависимости от флага `train_classifier`: с претренированной головой и со своей кастомной
> 
> 
> Претренированную голову точно можно ипользовать с помощью `CrossEncoder.predict()`, а о том что ее можно использовать и с помощью какого-то метода класса `SentenceTransformer` мне неизвестно. Если это и можно сделать, то наверное это что-то экзотическое и не задокументировано в самой библиотеке `sentence_transformers`

Да, нельзя использовать доп голову. Просто сейчас у нас реализован получение эмбеда из 0 токена (предположим, что это CLS) но не все модели обучались с ним и не факт, что это хорошо работать будет

**Comment by voorhs** [comment]
> > > Я говорю про то чтобы не использовать _classifier_hook можно просто сделать SentenceTranformer, а не CrossEncoder
> > 
> > 
> > причины почему мы используем хук:
> > ```
> > * это работает :)
> > 
> > * нам нужно использовать один трансформер в двух режимах в зависимости от флага `train_classifier`: с претренированной головой и со своей кастомной
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Претренированную голову точно можно ипользовать с помощью `CrossEncoder.predict()`, а о том что ее можно использовать и с помощью какого-то метода класса `SentenceTransformer` мне неизвестно. Если это и можно сделать, то наверное это что-то экзотическое и не задокументировано в самой библиотеке `sentence_transformers`
> 
> Да, нельзя использовать доп голову. Просто сейчас у нас реализован получение эмбеда из 0 токена (предположим, что это CLS) но не все модели обучались с ним и не факт, что это хорошо работать будет

согл, тогда оставим это на будущее, вынесем как отдельный аргумент способ пулинга активаций

а вообще я еще хочу потом сделать фичу чтобы можно было user-defined кросс-энкодер использовать

**Review by voorhs — APPROVED** [review]
